test/distributed/fsdp/test_fsdp_comm_hooks.py

```diff
@@ -0,0 +1,253 @@
+                self.assertEqual(entry.communication_hook, allreduce_hook.allreduce_hook)
+
+            dummy_state = DummyState(process_group=None)
+
+            if has_wrapping:
+                # Creating a list of non-root submodules to test
+                submodules = get_submodules(fsdp_model_with_hook)
+                # Check that assertion is raised for registering a comm hook on a non-root
+                with self.assertRaisesRegex(AssertionError, '^register_comm_hook can only be called on a root instance.$'):
+                    submodules[1].register_comm_hook(
+                        dummy_state,
+                        dummy_hook
+                    )
+
+                # Simmulate a registretion of a hook on a submodule
+                submodules[1]._hook_registered = True
+
+                # Check that an error is raised when some of submodules have a non-default hook assigned
+                with self.assertRaisesRegex(
+                    AssertionError,
+                    '^communication hook can be only registered once$'
+                ):
+                    fsdp_model_with_hook.register_comm_hook(
+                        dummy_state,
+                        dummy_hook
+                    )
+
+                # Reinitialize model to default
+                fsdp_model_with_hook = init_model(Net(has_wrapping))
+                submodules = get_submodules(fsdp_model_with_hook)
+                submodules[1].communication_hook = dummy_hook
+
+                # Check that an error is raised when some of submodules have a non-default hook assigned
+                with self.assertRaisesRegex(
+                    AssertionError,
+                    f'^communication hook should be default, but it is {submodules[1].communication_hook.__name__} instead$'
+                ):
+                    fsdp_model_with_hook.register_comm_hook(
+                        dummy_state,
+                        dummy_hook
+                    )
+
+                # Reinitialize model to default
+                fsdp_model_with_hook = init_model(Net(has_wrapping))
+
+            fsdp_model_with_hook.register_comm_hook(
+                dummy_state,
+                dummy_hook
+            )
+
+            # Check that we can't register comm hook twice
+            with self.assertRaisesRegex(AssertionError, '^communication hook can be only registered once$'):
+                fsdp_model_with_hook.register_comm_hook(
+                    dummy_state,
+                    dummy_hook
+                )
+
+            # Check dummy hook was registered for the root and all submodules if any
+            for entry in FSDP.fsdp_modules(fsdp_model_with_hook):
+                self.assertEqual(
+                    entry.communication_hook,
+                    dummy_hook
+                )
+                self.assertEqual(
+                    entry.communication_hook_state,
+                    dummy_state
+                )
```

Comment:
Something to consider: This one test `test_default_communication_hook_initialization` is testing quite a bit even though they may all be considered under the "hook registration interface": e.g. non-supported sharding strategies should error, submodule versus root module hook parity, and not registering a hook twice.

It could be worthwhile to break this test up (though it is up to you). If one assertion fails, I think pytest will terminate the test, which is why it is good to discretize unit tests. Of course, the granularity for discretization is subjective.

Reply:
Valid point. I've incorporated granularity into current tests.